src/routing/http-routing-v1.md

```diff
@@ -199,15 +218,37 @@ Access-Control-Allow-Origin: *
 Access-Control-Allow-Methods: GET, OPTIONS
 ```
 
+## Known Schemas
+
+This section contains a non-exhaustive list of known schemas that MAY be supported by clients and servers.
+
+### Peer
+
+The `peer` schema represents an arbitrary peer.
+
+```json
+{
+  "Schema": "peer",
+  "ID": "12D3K...",
+  "Addrs": ["/ip4/..."],
+  "Protocols": ["transport-bitswap", ...]
+}
+```
```

Comment:
@masih just a quick check-in around using this schema more for DHT proxying:

This schema is added because of `/routing/v1/peers` but we also want to use it on `/routing/v1/providers` in cases where:
- a single peer speaks multiple protocols and we know that upfront, and protocols have no special metadata like the filecoin one, or
-  we don't know the protocol, dont want to assume bitswap, but we know peerid and/or multiaddrs and want to return them,

To illustrate, the current response when a peer speaks more than one protocol:
```json
{
      "Protocol": "transport-bitswap",
      "Schema": "bitswap",
      "ID": "QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6ZmC",
      "Addrs": [
        "/dns4/elastic.dag.house/tcp/443/wss"
      ],
      "Metadata": "gBI="
    },
    {
      "Protocol": "unknown",
      "Schema": "unknown",
      "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
      "Addrs": [
        "/dns4/dag.w3s.link/tcp/443/https"
      ]
    },
```

Our plan for DHT proxy in Kubo ([wip PR](https://github.com/ipfs/kubo/pull/9877/)) is to not return `unknown` schema, nor assume `bitswap`.  

Instead, reuse  `peer` schema from this IPIP:

```json
{
      "Schema": "peer",
      "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
      "Addrs": [ 
        "/dns4/dag.w3s.link/tcp/443/https",
        "/dns4/elastic.dag.house/tcp/443/wss"
      ],
      "Protocols": [...] // optional entries, empty means unknown
    },
```

Any concerns with this approach?

Reply:
`Protocols` is an array of string labels correct? 

If so, this is different from the existing unknown (in [boxo](https://github.com/ipfs/boxo/blob/main/routing/http/types/provider_unknown.go#L46)) where one could have additional fields in a protocol to convey information o whoever understands that protocol/schema.  I suspect it would be limiting to only allow string lables in `Protocols` for `/providers`.

Reply:
Good catch @masih!

- yes, `Protocols` would be opaque strings
-  `Bytes` is undocumented, but sounds like we should add it here: good idea to allow opaque passtrough, it would also allow us to potentially pass signed records [mentioned](https://github.com/ipfs/specs/pull/417#issuecomment-1580795385) by @guillaumemichel 

My thinking is to add optional `ProtocolData` map, where keys would be the same as values in `Protocols`:

```json
{
      "Schema": "peer",
      "ID": "QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6Zmb",
      "Addrs": [ 
        "/dns4/example.com/tcp/443/foobar",
        "/dns4/example.com/tcp/443"
      ],
      "Protocols": [ "transport-bitswap", "transport-foobar", "signed-libp2p-peer-record" ],
      "ProtocolData": {
           "transport-foobar": "[base64-blob]",
           "signed-libp2p-peer-record": "[base64-blob]",
      },
    },
```

This way, client can easily check if data of some protocol is present, or ignore it otherwise.

Reply:
The rationale behind `Bytes` in the implementation is to, i believe, keep the raw content of the message for signature verification etc.  The way it currently works is that:
* all messages include `Protocol` and `Schema`. 
* For known protocols, the parser expects other well known fields.
* for unknown protocols, the parser simply dups the message in a field **at implementation level** for user to the parse as they see fit. 

This way, at spec level we don't need a dedicated `Bytes` field. We just expect JSON objects that have at least the fields they must have. The rest is left to the user. This IMHO is a much better UX, because the complexity can be encapsulated in implementation level without adding clause at spec level to say "If protocol is unknown check the `Bytes` field, if not etc. etc."

Reply:
@masih note  this IPIP is about Peer routing, not Content routing. 
Is your point that `Bytes` in `peer` Schema should not be a map, but a single item? 


I worry that a single value will not be enough: the IPFS ecosystem needs to have a surface for experimentation, we should allow alternative routing systems to have different types of signed provider records, which in my mind means this API should (at least in theory) allow passing the original bytes for more than one blob.

In practice, I don't expect us to use `Bytes` map much any time soon.

Or did I misread you and your point is that we should not have `peer` Schema at all, and instead return duplicated multiaddrs for every known transport protocol, and use `unknown` schema for unknown one?

Reply:
> Is your point that Bytes in peer Schema should not be a map, but a single item?

Hi @lidel, that's not my point. Let me rephrase my suggestion:

 * Flatten the `Bytes` field in the outer message, where the spec mechanism tolerates unknown fields.

For example, your earlier example would instead look something like this:

```json
{
  "Schema": "peer",
  "ID": "QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6Zmb",
  "Addrs": [
    "/dns4/example.com/tcp/443/foobar",
    "/dns4/example.com/tcp/443"
  ],
  "Protocols": [
    "transport-bitswap",
    "foobar",
    "signed-libp2p-peer-record"
  ],
  "foobar": "[base64-blob]",
  "signed-libp2p-peer-record": "[base64-blob]"
}
```

Here is another example that shows extensibility of the values where they don't have to be blobs:

```json
{
  "Schema": "peer",
  "ID": "QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6Zmb",
  "Addrs": [
    "/dns4/example.com/tcp/443/foobar",
    "/dns4/example.com/tcp/443"
  ],
  "Protocols": [
    "transport-bitswap",
    "foobar",
    "signed-libp2p-peer-record",
    "fish"
  ],
  "foobar": "[base64-blob]",
  "signed-libp2p-peer-record": "[base64-blob]",
  "fish": {
    "fresh-water": false,
    "name": "barreleye"
  }
}
```


A parser that does not know of "foobar" or "fish" protocol, would simply ignore the unknown fields.

Does that make sense? WDYT?

Reply:
Ack, then we are aligned! I don't feel strongly about `Bytes`, just want to document a way for doing this in the spec to remove guesswork. Putting additional fields in the outer message sgtm. 

Editorial tasks here would be (note to self)
- [ ] explicitly state that parser MUST allow for unknown fields and clients MUST ignore unknown ones
- [ ] add paragraph/example with `foobar` protocol and `[base64-blob]`

Reply:
@lidel I added that information. Feel free to edit/change.